test/Test/Pos/Ssc/GodTossing/ComputeSharesSpec.hs

```diff
@@ -44,13 +61,194 @@ spec = describe "computeSharesDistr" $ do
+    twoRichRichmenDesc = "Two richmen with almost half of stake, sum of distribution mustn't be big"
+    onePoorOneRichDesc = "Two richmen: first with mpcThd stake, second with remained stake,\
+    \ sum of distribution mustn't be big"
+    severalPoorRichmenDesc =
+        "Several poor richmen, sum of distribution mustn't be big"
+    severalPoorOneRichRichmenDesc =
+        "Several richmen, one of them control most of stake, sum of distribution mustn't be big"
+
+    severalSimilarRichmenDesc = "Richmen with stake distribution [0.24, 0.25, 0.26, 0.29]\
+    \ to test validity of @computeSharesDistr@"
+    twentyRichmen1Desc = "20 richmen with similar stake to test fairness of generated distribution"
+    twentyRichmen2Desc = "20 richmen with similar stake to test fairness of generated distribution"
+    validateFairnessDesc = "Given a valid richmen, validate fairness"
+
 computeShares' :: RichmenStakes -> Either TossVerFailure SharesDistribution
 computeShares' stake = computeSharesDistrPure stake genesisMpcThd
 
+mpcThd :: Double
+mpcThd = coinPortionToDouble genesisMpcThd
+
+-- Check that sum of distribtuion is less than passed value
+isLimitedBy :: Word16 -> SharesDistribution -> Bool
+isLimitedBy mx sd = sum (toList sd) <= mx
+
+-- This type descibes the
+-- (max inaccuracy, number of bad cases of the first type, number of bad cases of the second type)
+-- More datails see in the @isDistrFair@ description.
+type Unfairness = (Rational, Int, Int)
+
+-- We will call distribution is fair, if the max inaccuracy less than 0.05
+-- and number of bad cases less than 5% of all 2^n cases.
+-- Max inaccuracy is more or less heuristic value
+-- which means which inaccuracy we can get in the bad case.
+-- This inaccuracy can lead two bad situation:
+-- 1. when nodes mustn't reveal commitment, but they can
+-- 2. when nodes must reveal commitment, but they can't.
+-- We can get these situations when sum of stakes of nodes which sent shares close to 0.5.
+-- Max inaccuracy is comupted as difference between generated stake and 0.5.
+-- There is way to estimate inaccurasy as difference between real distribution and
+-- generated. On the one hand it make sense but on the other hand this difference
+-- isn't not so important.
+isDistrFair :: RichmenStakes -> SharesDistribution -> Bool
+isDistrFair rs sd
+    | length rs > 20 = error "Too many richmen"
+    | otherwise = do
+        let n = length rs
+        let totalCases = toRational (2::Int)^n :: Rational
+        let fivePerc = 0.05 :: Rational
+        let totalCoins = sum $ map unsafeGetCoin $ toList rs
+        let totalDistr = sum (toList sd)
+        let stakeholders = HM.keys rs
+        let distrs = map (findStk totalCoins totalDistr) stakeholders
+        let !(er, firstCase, secondCase) = fairBrute distrs 0 0
+        er < 0.05 &&
+            toRational firstCase < totalCases * fivePerc &&
+            toRational secondCase < totalCases * fivePerc
```

Comment:
Do we really need to check anything except `er < maxError` (i. e. are the last two conditions essential)?